File: server/core/src/main/java/dev/slimevr/posestreamer/BVHFileStream.java
package dev.slimevr.posestreamer;

import com.jme3.math.FastMath;
import dev.slimevr.tracking.processor.Bone;
import dev.slimevr.tracking.processor.skeleton.HumanSkeleton;
import io.github.axisangles.ktmath.Quaternion;
import io.github.axisangles.ktmath.Vector3;
import org.apache.commons.lang3.StringUtils;

import java.io.*;


public class BVHFileStream extends PoseDataStream {
	private static final int LONG_MAX_VALUE_DIGITS = Long.toString(Long.MAX_VALUE).length();

	private BVHSettings bvhSettings = BVHSettings.BLENDER;

	private final BufferedWriter writer;
	private long frameCount = 0;
	private long frameCountOffset;

	private float[] angleBuf = new float[3];

	public BVHFileStream(OutputStream outputStream) {
		super(outputStream);
		writer = new BufferedWriter(new OutputStreamWriter(outputStream), 4096);
	}

	public BVHFileStream(OutputStream outputStream, BVHSettings bvhSettings) {
		this(outputStream);
		this.bvhSettings = bvhSettings;
	}

	public BVHFileStream(File file) throws FileNotFoundException {
		super(file);
		writer = new BufferedWriter(new OutputStreamWriter(outputStream), 4096);
	}

	public BVHFileStream(File file, BVHSettings bvhSettings) throws FileNotFoundException {
		this(file);
		this.bvhSettings = bvhSettings;
	}

	public BVHFileStream(String file) throws FileNotFoundException {
		super(file);
		writer = new BufferedWriter(new OutputStreamWriter(outputStream), 4096);
	}

	public BVHFileStream(String file, BVHSettings bvhSettings) throws FileNotFoundException {
		this(file);
		this.bvhSettings = bvhSettings;
	}

	public BVHSettings getBvhSettings() {
		return bvhSettings;
	}

	public void setBvhSettings(BVHSettings bvhSettings) {
		this.bvhSettings = bvhSettings;
	}

	private String getBufferedFrameCount(long frameCount) {
		String frameString = Long.toString(frameCount);
		int bufferCount = LONG_MAX_VALUE_DIGITS - frameString.length();

		return bufferCount > 0 ? frameString + StringUtils.repeat(' ', bufferCount) : frameString;
	}

	private boolean isEndBone(Bone bone) {
		return bone == null || (!bvhSettings.shouldWriteEndNodes() && bone.getChildren().isEmpty());
	}

	private void writeBoneHierarchy(Bone bone) throws IOException {
		writeBoneHierarchy(bone, 0);
	}

	private void writeBoneHierarchy(Bone bone, int level) throws IOException {
		// Treat null as bone. This allows for simply writing empty end bones
		boolean isEndBone = isEndBone(bone);

		// Don't write end sites at populated bones, BVH parsers don't like that
		// Ex case caught: `joint{ joint{ end }, end, end }` outputs `joint{ end
		// }` instead
		// Ex case let through: `joint{ end }`
		boolean isSingleChild = bone == null
			|| bone.getParent() == null
			|| bone.getParent().getChildren().size() <= 1;
		if (isEndBone && !isSingleChild) {
			return;
		}

		String indentLevel = StringUtils.repeat("\t", level);
		String nextIndentLevel = indentLevel + "\t";

		// Handle ends
		if (isEndBone) {
			writer.write(indentLevel + "End Site\n");
		} else {
			writer.write((level > 0 ? indentLevel + "JOINT " : "ROOT ") + bone.getBoneType() + "\n");
		}
		writer.write(indentLevel + "{\n");

		// Ignore the root offset and original root offset
		if (level > 0 && bone != null && bone.getParent() != null) {
			float offsetScale = bvhSettings.getOffsetScale();
			writer
				.write(
					nextIndentLevel
						+ "OFFSET "
						+ 0
						+ " "
						+ -bone.getParent().getLength() * offsetScale
						+ " "
						+ 0
						+ "\n"
				);
		} else {
			writer.write(nextIndentLevel + "OFFSET 0.0 0.0 0.0\n");
		}

		// Handle ends
		if (!isEndBone) {
			// Only give position for root
			if (level > 0) {
				writer.write(nextIndentLevel + "CHANNELS 3 Zrotation Xrotation Yrotation\n");
			} else {
				writer
					.write(
						nextIndentLevel
							+ "CHANNELS 6 Xposition Yposition Zposition Zrotation Xrotation Yrotation\n"
					);
			}

			// If the bone has children
			if (!bone.getChildren().isEmpty()) {
				for (Bone childBone : bone.getChildren()) {
					writeBoneHierarchy(childBone, level + 1);
				}
			} else {
				// Write an empty end bone
				writeBoneHierarchy(null, level + 1);
			}
		}

		writer.write(indentLevel + "}\n");
	}

	@Override
	public void writeHeader(HumanSkeleton skeleton, PoseStreamer streamer) throws IOException {
		if (skeleton == null) {
			throw new NullPointerException("skeleton must not be null");
		}
		if (streamer == null) {
			throw new NullPointerException("streamer must not be null");
		}

		writer.write("HIERARCHY\n");
		writeBoneHierarchy(skeleton.getHeadBone());

		writer.write("MOTION\n");
		writer.write("Frames: ");

		// Get frame offset for finishing writing the file
		if (outputStream instanceof FileOutputStream fileOutputStream) {
			// Flush buffer to get proper offset
			writer.flush();
			frameCountOffset = fileOutputStream.getChannel().position();
		}

		writer.write(getBufferedFrameCount(frameCount) + "\n");

		// Frame time in seconds
		writer.write("Frame Time: " + (streamer.getFrameInterval() / 1000d) + "\n");
	}

	// Roughly based off code from
	// https://github.com/TrackLab/ViRe/blob/50a987eff4db31036b2ebaeb5a28983cd473f267/Assets/Scripts/BVH/BVHRecorder.cs
	private float[] quatToXyzAngles(Quaternion q, float[] angles) {
		if (angles == null) {
			angles = new float[3];
		} else if (angles.length != 3) {
			throw new IllegalArgumentException("Angles array must have three elements");
		}

		float x = q.getX();
		float y = q.getY();
		float z = q.getZ();
		float w = q.getW();

		// Roll (X)
		float sinrCosp = -2f * (x * y - w * z);
		float cosrCosp = w * w - x * x + y * y - z * z;
		angles[0] = FastMath.atan2(sinrCosp, cosrCosp);

		// Pitch (Y)
		float sinp = 2f * (y * z + w * x);
		// Use 90 degrees if out of range
		angles[1] = FastMath.abs(sinp) >= 1f
			? FastMath.copysign(FastMath.PI / 2f, sinp)
			: FastMath
				.asin(sinp);

		// Yaw (Z)
		float sinyCosp = -2f * (x * z - w * y);
		float cosyCosp = w * w - x * x - y * y + z * z;
		angles[2] = FastMath.atan2(sinyCosp, cosyCosp);

		return angles;
	}

	private void writeBoneHierarchyRotation(Bone bone, Quaternion inverseRootRot)
		throws IOException {
		Quaternion rot = bone.getGlobalRotation();

		// Adjust to local rotation
		if (inverseRootRot != null) {
			rot = inverseRootRot.times(rot);
		}

		// Roll (X), pitch (Y), yaw (Z) (intrinsic)
		// TODO: `quatToXyzAngles` should no longer be needed, this should be
		// changed to use `Quaternion.toEulerAngles()` instead
		angleBuf = quatToXyzAngles(rot.unit(), angleBuf);

		// Output in order of roll (Z), pitch (X), yaw (Y) (extrinsic)
		writer
			.write(
				angleBuf[0] * FastMath.RAD_TO_DEG
					+ " "
					+ angleBuf[1] * FastMath.RAD_TO_DEG
					+ " "
					+ angleBuf[2] * FastMath.RAD_TO_DEG
			);

		// Get inverse rotation for child local rotations
		if (!bone.getChildren().isEmpty()) {
			Quaternion inverseRot = bone.getGlobalRotation().inv();
			for (Bone childBode : bone.getChildren()) {
				if (isEndBone(childBode)) {
					// If it's an end bone, skip
					continue;
				}

				// Add spacing
				writer.write(" ");
				writeBoneHierarchyRotation(childBode, inverseRot);
			}
		}
	}

	@Override
	public void writeFrame(HumanSkeleton skeleton) throws IOException {
		if (skeleton == null) {
			throw new NullPointerException("skeleton must not be null");
		}

		Bone rootBone = skeleton.getHeadBone();

		Vector3 rootPos = rootBone.getPosition();

		// Write root position
		float positionScale = bvhSettings.getPositionScale();
		writer
			.write(
				rootPos.getX() * positionScale
					+ " "
					+ rootPos.getY() * positionScale
					+ " "
					+ rootPos.getZ() * positionScale
					+ " "
			);
		writeBoneHierarchyRotation(rootBone, null);

		writer.newLine();

		frameCount++;
	}

	@Override
	public void writeFooter(HumanSkeleton skeleton) throws IOException {
		// Write the final frame count for files
		if (outputStream instanceof FileOutputStream fileOutputStream) {
			// Flush before anything else
			writer.flush();
			// Seek to the count offset
			fileOutputStream.getChannel().position(frameCountOffset);
			// Overwrite the count with a new value
			writer.write(Long.toString(frameCount));
		}
	}

	@Override
	public void close() throws IOException {
		writer.close();
		super.close();
	}
}

File: server/core/src/main/java/dev/slimevr/posestreamer/BVHFileStream.kt
package dev.slimevr.posestreamer

import com.jme3.math.FastMath
import dev.slimevr.tracking.processor.Bone
import dev.slimevr.tracking.processor.skeleton.HumanSkeleton
import io.github.axisangles.ktmath.EulerOrder
import io.github.axisangles.ktmath.Quaternion
import io.github.axisangles.ktmath.Vector3
import org.apache.commons.lang3.StringUtils
import java.io.BufferedWriter
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream
import java.io.OutputStreamWriter

class BVHFileStream : PoseDataStream {
	var bvhSettings: BVHSettings = BVHSettings.BLENDER

	private val writer: BufferedWriter
	private var frameCount: Long = 0
	private var frameCountOffset: Long = 0

	constructor(outputStream: OutputStream, bvhSettings: BVHSettings = BVHSettings.BLENDER) : super(outputStream) {
		this.bvhSettings = bvhSettings
		writer = BufferedWriter(OutputStreamWriter(outputStream), 4096)
	}

	constructor(file: File, bvhSettings: BVHSettings = BVHSettings.BLENDER) : super(file) {
		this.bvhSettings = bvhSettings
		writer = BufferedWriter(OutputStreamWriter(outputStream), 4096)
	}

	constructor(file: String, bvhSettings: BVHSettings = BVHSettings.BLENDER) : super(file) {
		this.bvhSettings = bvhSettings
		writer = BufferedWriter(OutputStreamWriter(outputStream), 4096)
	}

	private fun getBufferedFrameCount(frameCount: Long): String {
		val frameString = frameCount.toString()
		val bufferCount = LONG_MAX_VALUE_DIGITS - frameString.length

		return if (bufferCount > 0) frameString + StringUtils.repeat(' ', bufferCount) else frameString
	}

	private fun internalNavigateSkeleton(
		bone: Bone,
		header: (bone: Bone, lastBone: Bone?, invertParentRot: Quaternion, distance: Int, hasBranch: Boolean, invertOffset: Boolean, zeroOffset: Boolean) -> Unit,
		footer: (distance: Int) -> Unit,
		lastBone: Bone? = null,
		invertParentRot: Quaternion = Quaternion.IDENTITY,
		distance: Int = 0,
		invertOffset: Boolean = false,
		zeroOffset: Boolean = false,
	) {
		val isRoot = lastBone == null
		val isInverse = bone === lastBone?.parent
		val parent = bone.parent

		// If we're visiting the parents or at root, continue to the next parent
		val visitParent = (isInverse || isRoot) && parent != null

		// We are navigating inversely; this bone's node is at the head rather than
		//  the tail, so we let our origin child take our children, and we take our
		//  parent's children in turn
		// Root is a special condition where it is both states simultaneously
		val invChildren = if (visitParent) {
			parent.children
		} else {
			emptyList()
		}
		val children = if (!isInverse) {
			bone.children
		} else {
			emptyList()
		}

		val hasBranch = visitParent || children.isNotEmpty()

		header(
			bone,
			lastBone,
			invertParentRot,
			distance,
			hasBranch,
			invertOffset,
			zeroOffset,
		)

		if (hasBranch) {
			// Cache this inverted rotation to reduce computation for each branch
			val thisInvertRot = bone.rotationOffset * bone.getGlobalRotation().inv()

			if (visitParent) {
				internalNavigateSkeleton(
					parent,
					header,
					footer,
					bone,
					thisInvertRot,
					distance + 1,
					invertOffset || isInverse,
					false,
				)
			}

			for (child in invChildren) {
				// Ignore our own bone (from parent's children)
				if (child == bone) continue
				internalNavigateSkeleton(
					child,
					header,
					footer,
					bone,
					thisInvertRot,
					distance + 1,
					true,
					!invertOffset,
				)
			}

			for (child in children) {
				internalNavigateSkeleton(
					child,
					header,
					footer,
					bone,
					thisInvertRot,
					distance + 1,
					false,
					invertOffset,
				)
			}
		}

		footer(distance)
	}

	private fun navigateSkeleton(
		root: Bone,
		header: (bone: Bone, lastBone: Bone?, invertParentRot: Quaternion, distance: Int, hasBranch: Boolean, invertOffset: Boolean, zeroOffset: Boolean) -> Unit,
		footer: (distance: Int) -> Unit = {},
		// Default true if the root isn't the true root (usually hip)
		tailRoot: Boolean = root.parent != null,
	) {
		// Root is treated as a parent if we want to target the tail as the root node
		internalNavigateSkeleton(root, header, footer, invertOffset = tailRoot)
	}

	private fun writeBoneDefHeader(bone: Bone?, lastBone: Bone?, invertParentRot: Quaternion, distance: Int, hasBranch: Boolean, invertOffset: Boolean, zeroOffset: Boolean) {
		val isRoot = lastBone == null
		val indentLevel = StringUtils.repeat("\t", distance)
		val nextIndentLevel = indentLevel + "\t"

		// Handle ends
		if (bone == null) {
			writer.write("${indentLevel}End Site\n")
		} else {
			writer.write("${indentLevel}${if (!isRoot) "JOINT" else "ROOT"} ${bone.boneType}\n")
		}
		writer.write("$indentLevel{\n")

		// "OFFSET": Defines the parent bone's local tail position
		// Ignore the root and endpoint offsets
		val offset = if (zeroOffset || isRoot) {
			Vector3.NULL
		} else {
			lastBone.rotationOffset.sandwichUnitY() * ((if (invertOffset) 1 else -1) * lastBone.length * bvhSettings.offsetScale)
		}
		writer.write("${nextIndentLevel}OFFSET ${offset.x} ${offset.y} ${offset.z}\n")

		// Define channels
		if (bone != null) {
			// Only give position for root
			if (!isRoot) {
				writer.write("${nextIndentLevel}CHANNELS 3 Zrotation Xrotation Yrotation\n")
			} else {
				writer.write(
					"${nextIndentLevel}CHANNELS 6 Xposition Yposition Zposition Zrotation Xrotation Yrotation\n",
				)
			}

			// Write an empty end bone if there are no branches
			// We use null for convenience and treat it as an end node (no bone)
			if (!hasBranch) {
				val endDistance = distance + 1
				writeBoneDefHeader(
					null,
					bone,
					Quaternion.IDENTITY,
					endDistance,
					hasBranch = false,
					invertOffset = false,
					zeroOffset = false,
				)
				writeBoneDefFooter(endDistance)
			}
		}
	}

	private fun writeBoneDefFooter(level: Int) {
		// Closing bracket
		writer.write("${StringUtils.repeat("\t", level)}}\n")
	}

	private fun writeSkeletonDef(rootBone: Bone) {
		navigateSkeleton(rootBone, ::writeBoneDefHeader, ::writeBoneDefFooter)
	}

	@Throws(IOException::class)
	override fun writeHeader(skeleton: HumanSkeleton, streamer: PoseStreamer) {
		writer.write("HIERARCHY\n")
		writeSkeletonDef(skeleton.getBone(bvhSettings.rootBone))

		writer.write("MOTION\n")
		writer.write("Frames: ")

		// Get frame offset for finishing writing the file
		if (outputStream is FileOutputStream) {
			// Flush buffer to get proper offset
			writer.flush()
			frameCountOffset = outputStream.channel.position()
		}

		writer.write(getBufferedFrameCount(frameCount) + "\n")

		// Frame time in seconds
		writer.write("Frame Time: ${streamer.frameInterval}\n")
	}

	private fun writeBoneRot(bone: Bone, lastBone: Bone?, invertParentRot: Quaternion, distance: Int, hasBranch: Boolean, invertOffset: Boolean, zeroOffset: Boolean) {
		val rot = bone.rotationOffset.inv() * invertParentRot * bone.getGlobalRotation()
		val angles = rot.toEulerAngles(EulerOrder.ZXY)

		// We're the root, so write position
		if (lastBone == null) {
			val rootPos = if (invertOffset) {
				bone.getTailPosition()
			} else {
				bone.getPosition()
			}
			// Write root position
			val positionScale = bvhSettings.positionScale
			writer.write("${rootPos.x * positionScale} ${rootPos.y * positionScale} ${rootPos.z * positionScale}")
		}

		// Output in order of roll (Z), pitch (X), yaw (Y) (extrinsic)
		// Assume spacing is needed at the start (we start with position with no following space)
		writer
			.write(" ${angles.z * FastMath.RAD_TO_DEG} ${angles.x * FastMath.RAD_TO_DEG} ${angles.y * FastMath.RAD_TO_DEG}")
	}

	@Throws(IOException::class)
	override fun writeFrame(skeleton: HumanSkeleton) {
		val rootBone = skeleton.getBone(bvhSettings.rootBone)
		navigateSkeleton(rootBone, ::writeBoneRot)
		writer.newLine()

		frameCount++
	}

	@Throws(IOException::class)
	override fun writeFooter(skeleton: HumanSkeleton) {
		// Write the final frame count for files
		if (outputStream is FileOutputStream) {
			// Flush before anything else
			writer.flush()
			// Seek to the count offset
			outputStream.channel.position(frameCountOffset)
			// Overwrite the count with a new value
			writer.write(frameCount.toString())
		}
	}

	@Throws(IOException::class)
	override fun close() {
		writer.close()
		super.close()
	}

	companion object {
		private const val LONG_MAX_VALUE_DIGITS = Long.MAX_VALUE.toString().length
	}
}
